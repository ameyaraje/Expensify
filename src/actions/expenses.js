import { v1 as uuidv1 } from 'uuid';

export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuidv1(),
            description: description,
            note: note,
            amount: amount,
            createdAt: createdAt
        }
    };
};

export const removeExpense = ({ id } = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    };
};

export const editExpense = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates  
    };
};