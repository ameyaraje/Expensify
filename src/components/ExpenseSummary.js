import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = (props) => {
    // console.log('-----> ', props);
    
    return (
        <div>
            {props.expensesLen === 1 ? (
                <p>Viewing 1 expense totalling {numeral(props.expenseTotal).format('$0,0.00')}</p>
            ) : (
                <p>Viewing {props.expensesLen} expenses totalling {numeral(props.expenseTotal).format('$0,0.00')}</p>
            )}
        </div>
    );
};

const mapStateToProps = (state, ownProps) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    const expenseTotal = getExpensesTotal(visibleExpenses);
    return {
        expensesLen: visibleExpenses.length,
        expenseTotal: expenseTotal
    };
};

export default connect(mapStateToProps)(ExpenseSummary);