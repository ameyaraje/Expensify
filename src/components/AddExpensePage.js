import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // console.log(expense);
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h2>
                    From the Add Expense Page
                </h2>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        
    )};
}


const mapDispatchToProps = (dispatch) => {
    return {
        startAddExpense: (expense) => {
            return dispatch(startAddExpense(expense));
        } 
    };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);