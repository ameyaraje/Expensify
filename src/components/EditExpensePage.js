import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => { 
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
        var id = this.props.expense.id;
        // console.log(id);
        this.props.removeExpense({ id });
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
            {/* 
                Editing expense with ID: {this.props.match.params.id}
            */}
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button 
                    onClick={this.onRemove} 
                >
                    Remove Expense
                </button>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => {
            return dispatch(editExpense(id, expense));
        },
        removeExpense: (data) => {
            return dispatch(removeExpense(data));
        }
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);