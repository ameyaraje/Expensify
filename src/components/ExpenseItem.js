import React from 'react';
import { Link } from 'react-router-dom';

import { removeExpense } from '../actions/expenses';

const ExpenseItem = (props) => {
    // console.log(props);
    return (
        <div>
            <p>
                Name: 
                <Link to={`/edit/${props.item.id}`}>
                    {props.item.description}
                </Link>
            </p>
            <p>
                Amount: {props.item.amount}
            </p>
            <p>
                CreatedAt: {props.item.createdAt}
            </p>
        </div>
    );
};

export default ExpenseItem;