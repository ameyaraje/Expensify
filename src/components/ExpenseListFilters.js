import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => {
            return {
                calendarFocused: calendarFocused
            };
        });
    };

    onFilterTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortByChange = (e) => {
        var filter = e.target.value;
        if (filter === 'amount') {
            console.log('sorting by amount');
            this.props.sortByAmount();
        }
        if (filter === 'date') {
            console.log('sorting by date');
            this.props.sortByDate();
        }
    };

    render() {
        return (
            <div>
                <p>Enter word to filter expense by here:</p>
                <input type="text" value={this.props.filters.text} onChange={this.onFilterTextChange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => (false)}
                    showClearDates={true}
                />
            </div>
        ); 
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => {
            return dispatch(setTextFilter(text));
        },
        sortByAmount: () => {
            return dispatch(sortByAmount());
        }, 
        sortByDate: () => {
            return dispatch(sortByDate());
        }, 
        setEndDate: (date) => {
            return dispatch(setEndDate(date))
        }, 
        setStartDate: (date) => {
            return dispatch(setStartDate(date))
        }
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);