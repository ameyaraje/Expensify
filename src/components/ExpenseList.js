import React from 'react';
import { connect } from 'react-redux';

import ExpenseItem from './ExpenseItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => {
    // console.log('------>', props.expenses[0]);
    
    return (
        <div>
            <h2>Here are all the expenses</h2>
            {
                props.expenses.length === 0 ? (
                    <h4>No Expenses to Show</h4>
                ) : (
                    props.expenses.map((expense) => {
                        return (<ExpenseItem key={expense.id} item={expense}/>);
                    })
                )
            }
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);