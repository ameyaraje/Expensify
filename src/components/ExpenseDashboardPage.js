import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => {
    return (
        <div>
            <h1>Expense Component</h1>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    );
};

export default ExpenseDashboardPage;