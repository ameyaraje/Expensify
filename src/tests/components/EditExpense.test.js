import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let editExpenseSpy, removeExpenseSpy, wrapper, historySpy;

beforeEach(() => {
    editExpenseSpy = jest.fn();
    historySpy = { push: jest.fn() };
    removeExpenseSpy = jest.fn();
    wrapper = shallow(<EditExpensePage 
        history={historySpy} 
        editExpense={editExpenseSpy} 
        removeExpense={removeExpenseSpy} 
        expense={expenses[0]}
    />);
});

test('should render EditExpense', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should edit an existing expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);

    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should remove an existing expense', () => {
    wrapper.find('button').simulate('click');

    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({ 
        id: expenses[0].id 
    });
});