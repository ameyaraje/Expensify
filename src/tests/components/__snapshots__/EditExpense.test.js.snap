// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpense 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 10,
        "createdAt": -345600000,
        "description": "bill 1",
        "id": "123",
        "note": "no note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
